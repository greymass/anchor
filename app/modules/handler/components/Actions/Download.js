// @flow
import React, { Component } from 'react';
import { withTranslation } from 'react-i18next';
import { Button, Header, Icon, Label } from 'semantic-ui-react';

const { clipboard, ipcRenderer } = require('electron');

class PromptActionDownload extends Component<Props> {
  onSaveUnsigned = () => {
    const { prompt, settings } = this.props;
    const { contract, resolved } = prompt;
    const { transaction } = resolved;
    const data = JSON.stringify({
      contract,
      transaction: {
        transaction_id: '',
        broadcast: false,
        transaction: {
          compression: 'none',
          signatures: [],
          transaction
        }
      }
    }, null, 2);
    ipcRenderer.send('saveFile', settings.lastFilePath, data);
  }
  render() {
    const {
      disabled,
      loading,
      t,
      wallet,
    } = this.props;
    return (
      <Button
        animated="fade"
        as="div"
        disabled={disabled}
        floated="right"
        labelPosition="left"
        onClick={this.onSaveUnsigned}
      >
        <Label
          as="a"
          basic
          pointing="right"
        >
          <Header textAlign="left">
            <Header.Content>
              <Header.Subheader style={{ fontWeight: 'bold' }}>
                {t('handler_action_download_header')}
              </Header.Subheader>
              {wallet.account}@{wallet.authorization}
            </Header.Content>
          </Header>
        </Label>
        <Button
          color="orange"
          icon
          loading={loading}
          style={{ position: 'relative' }}
        >
          <Button.Content hidden style={{ marginTop: '-0.75em' }}>
            <Icon
              name="save"
              size="large"
              style={{ marginRight: 0 }}
            />
          </Button.Content>
          <Button.Content visible style={{ margin: '0 1em' }}>
            <Icon
              name="save outline"
              size="large"
              style={{ marginRight: 0 }}
            />
          </Button.Content>
        </Button>
      </Button>
    );
  }
}

export default withTranslation('handler')(PromptActionDownload);
