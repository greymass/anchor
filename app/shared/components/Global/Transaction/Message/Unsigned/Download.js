// @flow
import React, { Component } from 'react';
import { withTranslation } from 'react-i18next';
import { Button, Divider, Grid, Header, Icon, Message, Modal, Segment, Tab } from 'semantic-ui-react';
import ReactJson from 'react-json-view';
import QRCode from 'qrcode';

import GlobalDangerLink from '../../../../../containers/Global/DangerLink';

const zlib = require('zlib');
const { clipboard, ipcRenderer } = require('electron');
const { SigningRequest } = require('eosio-signing-request');

const opts = {
  zlib: {
    deflateRaw: (data) => new Uint8Array(zlib.deflateRawSync(Buffer.from(data))),
    inflateRaw: (data) => new Uint8Array(zlib.inflateRawSync(Buffer.from(data))),
  },
};

export class GlobalTransactionMessageUnsignedDownload extends Component<Props> {
  constructor(props) {
    super(props);
    this.state = {
      copiedTransaction: false,
      uri: false
    };
  }
  componentDidMount() {
    this.renderQRCode();
  }
  async renderQRCode() {
    const { settings, transaction } = this.props;
    const { chainId } = settings;
    const req = await SigningRequest.create({
      chainId,
      transaction: transaction.transaction.transaction.transaction
    }, opts);
    const uri = req.encode();
    this.setState({ uri });
    const { canvas } = this;
    QRCode.toCanvas(canvas, uri, {
      scale: 6
    }, (error) => {
      if (error) console.error(error);
    });
  }
  promptSave = () => {
    const { contract, settings, transaction } = this.props;
    const data = JSON.stringify({
      contracts: transaction.contracts,
      transaction: transaction.transaction,
    }, null, 2);
    ipcRenderer.send('saveFile', settings.lastFilePath, data);
  };
  onCopy = () => {
    const {
      transaction
    } = this.props;
    const data = JSON.stringify({
      contracts: transaction.contracts,
      transaction: transaction.transaction,
    }, null, 2);
    this.setState({ copiedTransaction: true }, () => {
      clipboard.writeText(data);
      setTimeout(() => { this.setState({ copiedTransaction: false }); }, 5000);
    });
  };
  onTabChange = (e, { activeIndex }) => {
    if (activeIndex === 0) {
      this.renderQRCode();
    }
  }
  render() {
    const {
      onClose,
      t,
      transaction
    } = this.props;
    const {
      copiedTransaction,
      uri
    } = this.state;
    let uriParts = [];
    if (uri) {
      uriParts = uri.split('://');
    }
    const panes = [
      {
        menuItem: 'Signing',
        render: () => (
          <Tab.Pane>
            <Grid divided relaxed>
              <Grid.Row>
                <Grid.Column width={6}>
                  <Header
                    content="Export this Transaction"
                    subheader="Use one of the following options to export this unsigned transaction for use in another program or on another computer."
                    textAlign="center"
                  />
                  <Button
                    basic
                    color="blue"
                    content="Save as File"
                    fluid
                    icon="download"
                    onClick={this.promptSave}
                    style={{ marginBottom: '1em' }}
                  />
                  <Button
                    basic
                    color="blue"
                    content="Copy to Clipboard"
                    fluid
                    icon={(copiedTransaction) ? 'circle check' : 'clipboard'}
                    onClick={this.onCopy}
                    style={{ marginBottom: '1em' }}
                  />
                  <GlobalDangerLink
                    content={(
                      <Button
                        basic
                        color="blue"
                        content="View ESR Link"
                        fluid
                        icon="external"
                      />
                    )}
                    link={`https://eosio.to/${uriParts[1]}`}
                  />
                  <Message
                    icon
                    textAlign="left"
                  >
                    <Message.Content>
                      <Message.Header>
                        Expiration of Transactions
                      </Message.Header>
                      Every transaction has an expiration date, after which time even if signed will not be accepted by the blockchain. To successfully complete this transaction, ensureÂ it is signed within the given amount of time.
                    </Message.Content>
                  </Message>
                  {(copiedTransaction) && (
                    <Message
                      color="teal"
                      content={t('global_transaction_unsigned_copied')}
                      icon="check"
                    />
                  )}
                </Grid.Column>
                <Grid.Column width={10} textAlign="center">
                  <canvas ref={(node) => { this.canvas = node; }} />
                  <Header
                    content="Sign with a Mobile Wallet"
                    style={{ marginTop: 0 }}
                    subheader="Scan this QR Code with an ESR/EEP-7 compatible wallet to sign and complete this transaction."
                  />
                </Grid.Column>
              </Grid.Row>
            </Grid>
          </Tab.Pane>
        )
      },
      {
        menuItem: 'Transaction',
        render: () => (
          <Tab.Pane>
            <ReactJson
              collapsed={4}
              displayDataTypes={false}
              displayObjectSize={false}
              iconStyle="square"
              name={null}
              src={transaction}
              style={{ padding: '1em' }}
              theme="harmonic"
            />
          </Tab.Pane>
        )
      },
    ];
    return (
      <React.Fragment>
        <Header
          content="Unsigned Transaction"
          icon="qrcode"
          subheader="This read only wallet has created an unsigned transaction which needs to be signed by the appropriate account before broadcasting to the blockchain."
        />
        <Modal.Content>
          <Tab
            onTabChange={this.onTabChange}
            panes={panes}
          />
        </Modal.Content>
        <Modal.Actions>
          <Segment basic clearing>
            <Button color="green" floated="right" onClick={onClose}>
              <Icon name="checkmark" /> {t('close')}
            </Button>
          </Segment>
        </Modal.Actions>
      </React.Fragment>
    );
  }
}

export default withTranslation('global')(GlobalTransactionMessageUnsignedDownload);
