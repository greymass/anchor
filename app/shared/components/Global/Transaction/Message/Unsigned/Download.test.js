import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import { GlobalTransactionMessageUnsignedDownload } from './Download';

const mockProps = {
  onClose: () => null,
  t: text => text,
  transaction: {},
};

describe('GlobalTransactionMessageUnsignedDownload', () => {
  it('renders correctly', () => {
    const wrapper = shallow(<GlobalTransactionMessageUnsignedDownload {...mockProps} />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
