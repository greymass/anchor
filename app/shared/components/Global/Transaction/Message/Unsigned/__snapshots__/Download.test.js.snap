// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GlobalTransactionMessageUnsignedDownload renders correctly 1`] = `
<Segment
  basic={true}
>
  <Header
    content="global_transaction_unsigned_title"
    icon="checkmark"
    size="large"
  />
  <ModalContent>
    <p>
      global_transaction_unsigned_message
    </p>
    <Segment
      padded={true}
    >
      <Header
        content="global_transaction_unsigned_raw_tx"
      />
      <t
        collapseStringsAfterLength={false}
        collapsed={1}
        defaultValue={null}
        displayDataTypes={false}
        displayObjectSize={false}
        enableClipboard={true}
        groupArraysAfterLength={100}
        iconStyle="square"
        indentWidth={4}
        name={null}
        onAdd={false}
        onDelete={false}
        onEdit={false}
        onSelect={false}
        shouldCollapse={false}
        sortKeys={false}
        src={Object {}}
        style={
          Object {
            "padding": "1em",
          }
        }
        theme="harmonic"
        validationMessage="Validation Error"
      />
      <Segment
        basic={true}
        textAlign="center"
      >
        <Button
          as="button"
          color="blue"
          content="global_transaction_unsigned_save_file"
          icon="download"
          onClick={[Function]}
          style={
            Object {
              "marginBottom": "1em",
            }
          }
        />
        <Button
          as="button"
          color="teal"
          content="global_transaction_unsigned_copy_to_clipboard"
          icon="clipboard"
          onClick={[Function]}
        />
      </Segment>
    </Segment>
    <Message
      icon={true}
      size="large"
      warning={true}
    >
      <Icon
        as="i"
        name="info circle"
      />
      <MessageContent>
        <MessageHeader>
          global_transaction_unsigned_warning_title
        </MessageHeader>
        global_transaction_unsigned_warning_message
      </MessageContent>
    </Message>
  </ModalContent>
  <ModalActions>
    <Segment
      basic={true}
      clearing={true}
    >
      <Button
        as="button"
        color="green"
        floated="right"
        onClick={[Function]}
      >
        <Icon
          as="i"
          name="checkmark"
        />
         
        close
      </Button>
    </Segment>
  </ModalActions>
</Segment>
`;
